### PR TITLE
Support read-only cache

### PR DESCRIPTION
**Pull request by MarcoPolo**
In draft until I try this branch out somewhere

### CODE REVIEW

**Review by marten-seemann — COMMENTED** [review]
What's the use case for this?

**Comment by MarcoPolo** [comment]
Other implementations can benefit without having write access to the cache bucket (which would be a security concern).

**Comment by marten-seemann** [comment]
> Other implementations

Who specifically would use this? Is this for @thomaseizinger who's clearing his caches (too) frequently, or are there any other users?

**Comment by MarcoPolo** [comment]
zig-libp2p is using this. This will also really help nim-libp2p folks since a recent build of there's took ~30min: https://github.com/status-im/nim-libp2p/actions/runs/4385644603/jobs/7678638102.

At some point without caching the runners would run out of space. Building all these things takes a bit of a space.

**Comment by MarcoPolo** [comment]
cc @Menduist . This will speed up these jobs by about 20min. Ping me on slack and I'll share the read only keys :)

**Comment by marten-seemann** [comment]
> zig-libp2p is using this. This will also really help nim-libp2p folks since a recent build of there's took ~30min: https://github.com/status-im/nim-libp2p/actions/runs/4385644603/jobs/7678638102.
> 
> At some point without caching the runners would run out of space. Building all these things takes a bit of a space.

Thanks! Sounds good to me, let's ship it!

**Review by marten-seemann — APPROVED** [review]

**Review by thomaseizinger — COMMENTED** [review]
Thanks @MarcoPolo , this has been on my list for a while and I didn't get to it.

I did envision it slightly differently though. At the moment, CI runs from forks of our repos don't have access to the secrets and thus don't have access to the caches, right?

I figured instead of opting into a read-only cache, we should make that the default and configure the read-only keys as regular variables in our CI to allow forks access to it.

Writing to the cache should then be conditional on the actual secret being present. What do you think?

File: multidim-interop/src/generator.ts
Comment (MarcoPolo) [reply multidim-interop/src/generator.ts]: No, rolled it back

**Comment by MarcoPolo** [comment]
> I figured instead of opting into a read-only cache, we should make that the default and configure the read-only keys as regular variables in our CI to allow forks access to it.

True. I had some issue with this, but I'd like this to work as well. I believe the bucket is public to list, but not to get. I'll try to get this working.

**Comment by MarcoPolo** [comment]
Blocked on:
- [ ] https://github.com/libp2p/test-plans/issues/132#issuecomment-1466659605

**Comment by MarcoPolo** [comment]
Ugh, unfortunately you cannot simply omit the credentials. The AWS sdk wants the credentials to make signed requests to S3 (see the note [here](https://docs.aws.amazon.com/AmazonS3/latest/API/API_GetObject.html)). I was able to create a new IAM user with the following permissions on my personal account and have it work:

```
{
    "Version": "2012-10-17",
    "Statement": [
        {
            "Effect": "Allow",
            "Action": [
                "s3:Get*",
                "s3:List*"
            ],
            "Resource": "arn:aws:s3:::libp2p-by-tf-aws-bootstrap/*"
        }
    ]
}
```

Not as nice as seamlessly working with 0 setup. But at least it can be configured by individuals.

**Comment by thomaseizinger** [comment]
That is kind of annoying because requests via HTTP work without any setup?

**Comment by MarcoPolo** [comment]
Yup!

**Comment by galargh** [comment]
This will allow using anonymous creds for public buckets: https://github.com/moby/buildkit/pull/3692

**Comment by thomaseizinger** [comment]
> Ugh, unfortunately you cannot simply omit the credentials. The AWS sdk wants the credentials to make signed requests to S3 (see the note [here](https://docs.aws.amazon.com/AmazonS3/latest/API/API_GetObject.html)). I was able to create a new IAM user with the following permissions on my personal account and have it work:
> 
> ```
> {
>     "Version": "2012-10-17",
>     "Statement": [
>         {
>             "Effect": "Allow",
>             "Action": [
>                 "s3:Get*",
>                 "s3:List*"
>             ],
>             "Resource": "arn:aws:s3:::libp2p-by-tf-aws-bootstrap/*"
>         }
>     ]
> }
> ```
> 
> Not as nice as seamlessly working with 0 setup. But at least it can be configured by individuals.

Can we create a set of credentials that we default to?

To properly make this work for forks, we need it to be the same workflow, i.e. we can't specify a flag on the action whether we want to read or read-write.

Instead, I'd suggest that we default to a read-only user in the `dockerBuildWrapper.sh` script. That would also allow us to use this cache locally. In case a user is specified via the env variables, we override it and expect the credentials to have read-write permissions.

Thoughts?

**Comment by thomaseizinger** [comment]
> Personally, I'd wait for [moby/buildkit#3692](https://github.com/moby/buildkit/pull/3692) to be merged and then I'd start using the buildkit version from master. It's already approved so hopefully it's not long until it's merged.

Sounds good.

**Comment by MarcoPolo** [comment]
With #165 this is done

**Pull request closed without merging**
